src/dbnode/persist/fs/persist_manager.go

```diff
@@ -598,20 +603,25 @@ func (pm *persistManager) doneShared() error {
 	return nil
 }
 
-func (pm *persistManager) dataFilesetExistsAt(prepareOpts persist.DataPrepareOptions) (bool, error) {
+func (pm *persistManager) dataFilesetExists(prepareOpts persist.DataPrepareOptions) (bool, error) {
 	var (
-		blockStart = prepareOpts.BlockStart
-		shard      = prepareOpts.Shard
 		nsID       = prepareOpts.NamespaceMetadata.ID()
+		shard      = prepareOpts.Shard
+		blockStart = prepareOpts.BlockStart
+		volume     = prepareOpts.VolumeIndex
 	)
 
 	switch prepareOpts.FileSetType {
 	case persist.FileSetSnapshotType:
-		// Snapshot files are indexed (multiple per block-start), so checking if the file
-		// already exist doesn't make much sense
+		// Checking if a snapshot file exists for a block start doesn't make
+		// sense in this context because the logic for creating new snapshot
+		// files does not use the volume index provided in the prepareOpts.
+		// Instead, the new volume index is determined by looking at what files
+		// exist on disk. This means that there can never be a conflict when
+		// trying to write new snapshot files.
```

Comment:
Kind of strange, but why does snapshots use one mechanism (i.e. increment new volume index by looking at files on disk) vs fileset volumes using a different one?  Richie alluded that it was "hard" but would be interesting to hear what was the difficulty with just making them use similar/same code paths?

Reply:
Yeah I kind of did a poor job explaining this. There are a couple of subtle complexities that we encountered but basically we realized that we need to maintain some kind of relationship between the flush version and the volume number for a variety of reasons but the one that immediately comes to mind is imagine the `SeekerManager` does a lot of checks to see if specific fileset files exist. In order to do that efficiently it needs to be able to go from namespace/shard/blockStart/volume -> expected filename.

Once we decided that we wanted to maintain that relationship the "simplest" thing is to ensure that the flush version and the volume number are always the same. Technically we could do the `+1` logic and still maintain this guarantee but that would make it implicit instead of explicit and a brittle assumption.

So we decided to just tell the `PersistManager` / writer which volume we intended to create and that has some nice properties like if our logic is wrong the `PersistManager` can error out and tell us that the volume already exists (as opposed to it just incrementing by 1 and subtly breaking our assumptions).

Sine we took this route we do need to bootstrap the flush version back into the shard, but thats a trivial change since the shard already calls `ForEachInfoFile` so all we need to do is pull the volume number out of the info metadata and into the shard memory state.

@justinjc Do you remember any other details? We spent a few hours working through all this and I feel like I may be missing some of the other reasons we couldn't do the incrementing.

@robskillington Thinking about this now though it seems like it might be good to add a check in the persist manager that the volume being writeen is at least as high as the highest existing complete fileset file. I.E if complete versions of 1, 2, 3 exist then writing 3 (if `deleteIfExist=true`) and 4 should be allowed, but writing 2 should not. What do you think @justinjc ? Could also add it to your list of follow ups

Reply:
We could also support a flag in the `PersistManager` that basically tells it to figure out the volume number itself which we would probably need if we wanted to allow incremental peer bootstrapping and flushing to occur in parallel, but not sure we want to go with that approach long term (see my other comment below)

Reply:
@justinjc just pinging your thoughts on that potential follow up

Reply:
Thanks for the explanation here @richardartoul, you summed it up pretty well. In short, we wanted an efficient way to figure out what is the correct volume for a particular namespace/shard/block, and having to go to disk, parsing filenames and then ordering them each time seemed significantly slower.

I'm not a big fan of the persist manager having the volume number check. The check would similarly involve looking for files in the specified directory, doing file name parsing, and ordering them in order to figure that out, which amounts to additional logic that's essentially double checking that other logic isn't buggy. Also, from a separation of concerns point of view, I don't think the persist manager should need to know that fileset volumes are always increasing. Its job should just be to persist things you tell it to persist. It should be caller of the persist manager that ensures the volume number it passes is correct.